rust/lance-linalg/src/kmeans.rs

```diff
@@ -190,64 +185,38 @@ pub struct KMeanMembership {
 impl KMeanMembership {
     /// Reconstruct a KMeans model from the membership.
     async fn to_kmeans(&self) -> Result<KMeans> {
+        let time = std::time::Instant::now();
         let dimension = self.dimension;
-        let cluster_ids = Arc::new(self.cluster_ids.clone());
 
-        // New centroids for each cluster
-        let means = stream::iter(0..self.k)
-            .zip(repeat_with(|| {
-                (
-                    self.data.clone(),
-                    cluster_ids.clone(),
-                    self.centroids.clone(),
-                )
-            }))
-            .map(
-                |(cluster, (data, cluster_ids, prev_centroids))| async move {
-                    tokio::task::spawn_blocking(move || {
-                        let mut sum = vec![0.0; dimension];
-                        let data = data.values();
-                        let mut total = 0.0;
-                        // Eager group-by cluster id.
-                        for i in 0..cluster_ids.len() {
-                            if cluster_ids[i] as usize == cluster {
-                                // TODO: use simd ADD
-                                for j in 0..dimension {
-                                    sum[j] += data[i * dimension + j];
-                                }
-                                total += 1.0;
-                            };
-                        }
-                        if total > 0.0 {
-                            let s = Float32Array::from(
-                                sum
-                            );
-                            s.unary_mut(|x| x / total).unwrap()
-                        } else {
-                            warn!("Warning: KMean: cluster {} has no value, does not change centroids.", cluster);
-                            prev_centroids.slice(cluster * dimension, dimension)
-                        }
-                    })
-                    .await
-                },
-            )
-            .buffered(num_cpus::get())
-            .try_collect::<Vec<_>>()
-            .await.map_err(|e| {
-                ArrowError::ComputeError(format!(
-                    "KMeans: failed to compute new centroids: {}",
-                    e
-                ))
-            })?;
-
-        // TODO: concat requires `&[&dyn Array]`. Are there cheaper way to pass Vec<Float32Array> to `concat`?
-        let mut mean_refs: Vec<&dyn Array> = vec![];
-        for m in means.iter() {
-            mean_refs.push(m);
+        let mut cluster_cnts = vec![0_usize; self.k];
+        let mut new_centroids = vec![0.0_f32; self.k * dimension];
+        self.data
+            .values()
+            .chunks_exact(dimension)
+            .zip(self.cluster_ids.iter())
+            .for_each(|(v, c)| {
+                cluster_cnts[*c as usize] += 1;
+                for i in 0..dimension {
+                    new_centroids[*c as usize * dimension + i] += v[i];
+                }
```

Comment:
This might be a bit faster (especially for large `dimension`) as it can avoid some bounds checking.

```suggestion
            for (old, new) in new_centroids[(*c * dimension)..((*c + 1) * dimension)]
                .iter_mut()
                .zip(v)
            {
                *old += new;
            }
```

Reply:
Done